rfc/rfc-83/rfc-83.md

```diff
@@ -0,0 +1,260 @@
+{"name": "earliestCommitToRetain", "type": "string"}
+```
+
+We also need a unified interface/abstract-class to control the Plan behavior and Commit behavior of the TableService.
+
+### Abstraction
+
+Use `TableServiceCommitter` to control table service related commit action.
+
+```java
+package org.apache.hudi.table.action;
+
+public interface TableServiceCommitter<R> {
+
+  /**
+   * Build table service related commit metadata, 
+   * Especially `EarliestCommitToReta` column when enable incremental table service.
+   * @return metadata to commit
+   */
+  R buildCommitMeta();
+}
+```
+
+Use `TableServiceBaseActionExecutor` to control the behavior of getting partitions to be performed into current table service.
+
+```java
+package org.apache.hudi.table.action;
+
+import org.apache.hudi.common.engine.HoodieEngineContext;
+import org.apache.hudi.common.model.TableServiceType;
+import org.apache.hudi.common.table.timeline.HoodieInstant;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.config.HoodieWriteConfig;
+import org.apache.hudi.exception.HoodieNotSupportedException;
+import org.apache.hudi.table.HoodieTable;
+
+import java.io.IOException;
+import java.util.List;
+
+public abstract class TableServiceBaseActionExecutor<T, I, K, O, R> extends BaseActionExecutor<T, I, K, O, R> implements TableServiceCommitter<R> {
+
+  public TableServiceBaseActionExecutor(HoodieEngineContext context, HoodieWriteConfig config, HoodieTable<T, I, K, O> table, String instantTime) {
+    super(context, config, table, instantTime);
+  }
+
+  /**
+   * Returns the earliest commit to retain and will record in meta as new EarliestCommitToRetain
+   */
+  public Option<HoodieInstant> getEarliestCommitToRetain() {
+    return null;
+  }
+
+  /**
+   * get partitions to be performed in current table service
+   * @param instantToRetain
+   * @param type
+   * @return
+   * @throws IOException
+   */
+  public List<String> getPartitionPaths(Option<HoodieInstant> instantToRetain, TableServiceType type) {
+    return null;
+  }
+
+  /**
+   * Get incremental partitions from EarliestCommitToRetain to instantToRetain
+   * @param instantToRetain
+   * @param type
+   * @param deleteEmptyCommit
+   * @return
+   * @throws IOException
+   */
+  private List<String> getIncrementalPartitionPaths(Option<HoodieInstant> instantToRetain, TableServiceType type) {
+    return null;
+  }
+
+  /**
+   * Get all table partitions related to current table
+   * @param instantToRetain
+   * @param type
+   * @param deleteEmptyCommit
+   * @return
+   * @throws IOException
+   */
+  private List<String> getAllPartitionPaths(Option<HoodieInstant> instantToRetain, TableServiceType type) {
+    return null;
+  }
+
+  public R buildCommitMeta() {
+    return null;
+  }
+
+}
+
+```
+
+### Optimize the inheritance relationship of Table Service
+
+![NewTableServiceDiagram.png](NewTableServiceDiagram.png)
+
+### Work Flow for Incremental Table Service
+
+Table Service Planner
```

Comment:
Currently we are using `completion time`, should we indicate here whether `instant` refers to `request time` or `completion time`?

Reply:
Request time I believe. If using completion, it may miss some instant when multi-write.